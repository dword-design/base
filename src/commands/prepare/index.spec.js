import {
  endent,
  identity,
  includes,
  keyBy,
  mapValues,
  stubTrue,
} from '@dword-design/functions'
import tester from '@dword-design/tester'
import testerPluginTmpDir from '@dword-design/tester-plugin-tmp-dir'
import execa from 'execa'
import fs from 'fs-extra'
import globby from 'globby'
import outputFiles from 'output-files'

import { Base } from '@/src/index.js'

export default tester(
  {
    'additional allowed match': async () => {
      await fs.outputFile('foo.txt', '')

      const base = new Base({ allowedMatches: ['foo.txt'] })
      await base.prepare()
      expect(
        globby('*', { dot: true }) |> await |> includes('foo.txt')
      ).toBeTruthy()
    },
    'commit valid': async () => {
      await execa.command('git init')
      await execa.command('git config user.email "foo@bar.de"')
      await execa.command('git config user.name "foo"')

      const base = new Base()
      await base.prepare()
      await execa('git', ['commit', '--allow-empty', '-m', 'fix: foo'])
    },
    'commit with linting errors': async () => {
      await execa.command('git init')
      await execa.command('git config user.email "foo@bar.de"')
      await execa.command('git config user.name "foo"')

      const base = new Base()
      await base.prepare()
      await expect(
        execa.command('git commit --allow-empty -m foo')
      ).rejects.toThrow('subject may not be empty')
    },
    'custom prepare': async () => {
      const base = new Base({ prepare: () => fs.outputFile('foo.txt', 'bar') })
      await base.prepare()
      expect(await fs.readFile('foo.txt', 'utf8')).toEqual('bar')
    },
    async valid() {
      await execa.command('git init')
      await execa.command(
        'git remote add origin git@github.com:dword-design/bar.git'
      )
      await outputFiles({
        '.env.json': '',
        '.env.schema.json': endent`
      {
        "foo": "bar"
      }
    `,
        '.test.env.json': '',
        'CHANGELOG.md': '',
        'package.json': JSON.stringify({
          author: 'dword-design',
          license: 'MIT',
          name: 'foo',
        }),
        'src/index.js': 'export default 1',
        'yarn.lock': '',
      })

      const base = new Base()
      await base.prepare()
      expect(
        globby('*', { dot: true, onlyFiles: false })
          |> await
          |> keyBy(identity)
          |> mapValues(stubTrue)
      ).toMatchSnapshot(this)
      expect(await fs.readFile('README.md', 'utf8')).toMatchSnapshot(this)
      expect(await fs.readFile('LICENSE.md', 'utf8')).toMatch('MIT License')
    },
  },
  [testerPluginTmpDir()]
)
