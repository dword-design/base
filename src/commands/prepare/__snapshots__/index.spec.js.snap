// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index valid 1`] = `
Object {
  ".babelrc.json": true,
  ".commitlintrc.json": true,
  ".cz.json": true,
  ".devcontainer": true,
  ".editorconfig": true,
  ".env.json": true,
  ".env.schema.json": true,
  ".eslintrc.json": true,
  ".git": true,
  ".gitattributes": true,
  ".github": true,
  ".gitignore": true,
  ".gitpod.Dockerfile": true,
  ".gitpod.yml": true,
  ".husky": true,
  ".releaserc.json": true,
  ".renovaterc.json": true,
  ".test.env.json": true,
  ".vscode": true,
  "CHANGELOG.md": true,
  "LICENSE.md": true,
  "README.md": true,
  "package.json": true,
  "src": true,
  "yarn.lock": true,
}
`;

exports[`index valid 2`] = `
"<!-- TITLE/ -->
# foo
<!-- /TITLE -->

<!-- BADGES/ -->
  <p>
    <a href=\\"https://npmjs.org/package/foo\\">
      <img
        src=\\"https://img.shields.io/npm/v/foo.svg\\"
        alt=\\"npm version\\"
      >
    </a><img src=\\"https://img.shields.io/badge/os-linux%20%7C%C2%A0macos%20%7C%C2%A0windows-blue\\" alt=\\"Linux macOS Windows compatible\\"><a href=\\"https://github.com/dword-design/bar/actions\\">
      <img
        src=\\"https://github.com/dword-design/bar/workflows/build/badge.svg\\"
        alt=\\"Build status\\"
      >
    </a><a href=\\"https://codecov.io/gh/dword-design/bar\\">
      <img
        src=\\"https://codecov.io/gh/dword-design/bar/branch/master/graph/badge.svg\\"
        alt=\\"Coverage status\\"
      >
    </a><a href=\\"https://david-dm.org/dword-design/bar\\">
      <img src=\\"https://img.shields.io/david/dword-design/bar\\" alt=\\"Dependency status\\">
    </a><img src=\\"https://img.shields.io/badge/renovate-enabled-brightgreen\\" alt=\\"Renovate enabled\\"><br/><a href=\\"https://gitpod.io/#https://github.com/dword-design/bar\\">
      <img
        src=\\"https://gitpod.io/button/open-in-gitpod.svg\\"
        alt=\\"Open in Gitpod\\"
        width=\\"114\\"
      >
    </a><a href=\\"https://www.buymeacoffee.com/dword\\">
      <img
        src=\\"https://www.buymeacoffee.com/assets/img/guidelines/download-assets-sm-2.svg\\"
        alt=\\"Buy Me a Coffee\\"
        width=\\"114\\"
      >
    </a><a href=\\"https://paypal.me/SebastianLandwehr\\">
      <img
        src=\\"https://sebastianlandwehr.com/images/paypal.svg\\"
        alt=\\"PayPal\\"
        width=\\"163\\"
      >
    </a><a href=\\"https://www.patreon.com/dworddesign\\">
      <img
        src=\\"https://sebastianlandwehr.com/images/patreon.svg\\"
        alt=\\"Patreon\\"
        width=\\"163\\"
      >
    </a>
</p>
<!-- /BADGES -->

<!-- DESCRIPTION/ -->

<!-- /DESCRIPTION -->

<!-- INSTALL/ -->
## Install

\`\`\`bash
# npm
$ npm install foo

# Yarn
$ yarn add foo
\`\`\`
<!-- /INSTALL -->

<!-- LICENSE/ -->
## Contribute

Are you missing something or want to contribute? Feel free to file an [issue](https://github.com/dword-design/bar/issues) or a [pull request](https://github.com/dword-design/bar/pulls)! ‚öôÔ∏è

## Support

Hey, I am Sebastian Landwehr, a freelance web developer, and I love developing web apps and open source packages. If you want to support me so that I can keep packages up to date and build more helpful tools, you can donate here:

<p>
  <a href=\\"https://www.buymeacoffee.com/dword\\">
    <img
      src=\\"https://www.buymeacoffee.com/assets/img/guidelines/download-assets-sm-2.svg\\"
      alt=\\"Buy Me a Coffee\\"
      width=\\"114\\"
    >
  </a>&nbsp;If you want to send me a one time donation. The coffee is pretty good üòä.<br/>
  <a href=\\"https://paypal.me/SebastianLandwehr\\">
    <img
      src=\\"https://sebastianlandwehr.com/images/paypal.svg\\"
      alt=\\"PayPal\\"
      width=\\"163\\"
    >
  </a>&nbsp;Also for one time donations if you like PayPal.<br/>
  <a href=\\"https://www.patreon.com/dworddesign\\">
    <img
      src=\\"https://sebastianlandwehr.com/images/patreon.svg\\"
      alt=\\"Patreon\\"
      width=\\"163\\"
    >
  </a>&nbsp;Here you can support me regularly, which is great so I can steadily work on projects.
</p>

Thanks a lot for your support! ‚ù§Ô∏è

## License

[MIT License](https://opensource.org/license/mit/) ¬© [Sebastian Landwehr](https://sebastianlandwehr.com)
<!-- /LICENSE -->
"
`;
